### DISCUSSION TITLE
[Question]: How do we debug what exactly gets sent to APIs configured as actions

### DISCUSSION THREAD

**Discussion by dividor** [danny-avila/LibreChat#2997] [Q&A]
### What is your question?

Librechat is given an error, coming from OpenAI, that the tool output is unexpected because it's empty. However, when calling the API, from within the Librechat docker container, the tool outputs are empty. 

### More Details

In Librechat, we get this error, suggesting we sent tool output to OpenAI and they were empty ...

`Something went wrong. Here's the specific error message we encountered: 400 Expected tool outputs for call_ids ['call_UMzqTOfk2toOKrKyvvdXBxhh'], got []`

It's almost instantaneous.

In the Librechat UI, we see the parameters for the tool API call ...

```
{
  "query": "select table_name from table_metadata where summary like '%Chad%'"
}
```

This works perfectly in the action's API endpoint. I can also confirm curl of the call from within the Librechat docker container works. fine. What's weird, is that this all used to work perfectly, but started failing this morning (we use the `FROM ghcr.io/danny-avila/librechat-dev:latest` Docker image, and have tried ghcr.io/danny-avila/librechat also. We tried with a vanilla environment, still getting the same issue.

Logs:

[logs.tar.gz](https://github.com/user-attachments/files/15742396/logs.tar.gz)

I would love to debug what Librechat sends to the tool (action) to see if there is something wrong there. 

- Do you have any tips to debug this please?
- Any thoughts on what the issue might be?

### What is the main subject of your question?

Other

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by danny-avila** [comment]
without your actual openapi schema, it's hard to say what the cause is. there could be an error with the API response.

**Reply by danny-avila** [reply]
or rather, how the params get formatted by the assistant to the action handling.

**Reply by dividor** [reply]
Hi Danny, attached below.

I think my question is more - how can I debug the calls? I agree the issue is likely how the query is formed, but I was wondering how I might debug it myself. 

Thanks!

[haa_datarecipes.yaml.zip](https://github.com/user-attachments/files/15750495/haa_datarecipes.yaml.zip)

**Reply by danny-avila** [reply]
Breakpoints and a Javascript debugger. VS Code has a great one, built-in that I use everyday. You need to develop over the backend without docker involved, and you can later build your own image if docker is needed: https://www.librechat.ai/docs/development/get_started
